ykllvmwrap/src/ykllvmwrap.cc

```diff
@@ -174,6 +174,91 @@ extern "C" void *compile_module(string TraceName, Module *M) {
   return (void *)EE->getFunctionAddress(TraceName);
 }
 
+#ifndef NDEBUG
+// Left trim (in-place) the character `C` from the string `S`.
+void lTrim(string &S, const char C) {
+  S.erase(0, std::min(S.find_first_not_of(C), S.size() - 1));
+}
+
+// Dumps an LLVM Value to a string and trims leading whitespace.
+void dumpValueToString(Value *V, string &S) {
+  raw_string_ostream RSO(S);
+  V->print(RSO);
+  lTrim(S, ' ');
+}
+
+// Dump an annotated version of the JITMod to stderr, showing how the
+// instructions in the JITMod were derived from AOTMod.
+void dumpAnnotatedTrace(Module *AOTMod, Module *JITMod,
+                        ValueToValueMapTy *VMap) {
+  assert(JITMod->size() == 1);
+  Function *JITFunc = &*JITMod->begin();
+
+  // Build the inverse of the VMap.
+  ValueToValueMapTy RevVMap;
+  for (auto const &Elem : *VMap) {
+    if (isa<Instruction>(Elem.first)) {
+      // Skip things which are not an instruction, e.g. function arguments.
+      RevVMap[Elem.second] = const_cast<Value *>(Elem.first);
+    }
+  }
+
+  // Find the longest instruction from the JITMod so that we can align the
+  // second column.
+  size_t LongestJITLine = 0;
+  for (auto &JITBlock : *JITFunc) {
+    for (auto &JITInst : JITBlock) {
+      string Line;
+      dumpValueToString(&JITInst, Line);
+      auto Len = Line.length();
+      if (Len > LongestJITLine)
+        LongestJITLine = Len;
+    }
+  }
+
+  const string JITHeader = string("JIT Module");
+  string Padding = string(LongestJITLine - JITHeader.length(), ' ');
+  errs() << "\n\n--- Begin annotated trace dump for " << JITFunc->getName()
```

Comment:
"annotated trace dump" feels a mouthful. Just "annotated trace" or even just "trace" maybe?

Reply:
"annotated" is to distinguish this from the other kind of trace dump that we already have (for use by lang_tester IR tests).

"annotated trace" works for me.

Reply:
I think "side-by-side" might be better ("annotated" is a very generic term that doesn't really say much)?

Reply:
I don't like side-by-side.

Annotated feels right here. We are marking (annotating) the inlining boundaries.

@ptersilie any thoughts?

Reply:
Won't we have other types of annotated trace things? Basically I'm saying that "annotated" and "dump" are not adding value here (but they are adding a (rather small, I admit) cognitive load).

Reply:
I'd still like to hear what Lukas has to say, but note that "dump" is a convention used in LLVM for printing debugging info the stderr. You can invoke `->dump()` on pretty much anything and get a textual version to eyeball.

Reply:
I'm not fussed about the name. Just call it `Begin trace dump`, then we don't need to worry about whether `annotated` or `side-by-side` is better.

Reply:
The problem is, we already have the other trace dump used by tests, so we need to differentiate them with some term.

Reply:
The other dump doesn't print any headers though. I don't think it's an issue.

Reply:
What env var should control this? Currently you turn it on with `YKDEBUG_PRINT_IR_ANNOTATED=1`.

The tests set `YKDEBUG_PRINT_IR=1`.

Reply:
Personally I think everything should be `YK_` and that `DEBUG_PRINT` just says the same thing twice. So I think something like `YK_PRINT_IR` and `YK_PRINT_SBS` for side-by-side might be right. I know I'm repeating myself, but I'm convinced we'll have other "annotated" print thingies, so "annotated" just doesn't feel like a useful discriminator.